### DISCUSSION TITLE
How do you convert open_agb_firm saves to mgba?

### DISCUSSION THREAD

**Discussion by kanjieater** [profi200/open_agb_firm#211] [Q&A]
Are there any tools or scripts that people use to make this easy?  If anyone is can document the steps I'd find it really helpful to know how to convert to and from the formats.

I know there's been a PR sitting out around this for a while https://github.com/profi200/open_agb_firm/pull/178.

**Comment by profi200** [comment]
A similar option already exists. But be warned that EEPROM saves won't work as is. There is a very high chance games will try to delete your save if you use games with EEPROM saves in mGBA and oaf at the same time.

https://github.com/profi200/open_agb_firm/commit/877f7c61d078e6d2fa76aabdbf2ba229f9bb5928
And see the README reagrding EEPROM save conversion.

**Reply by kanjieater** [reply]
Thank you. I thought I had tried the eeprom converter to get saves from oaf back to mgba and it hadn't work. Undoubtedly I need to try again, since as you pointed out, it's in the readme, and pretty simple code. Thanks for pointing it out to me. UPDATE: Must have been user error before. Working fine for me

Could you talk a bit more about why  the eeprom save wouldn't work / be deleted for using between oaf & mgba?

I am planning to use mgba via retroarch on a PC, so not directly inside of the 3ds, so I'm wondering if the "deletion" scenario would apply if I'm just manually moving saves back and forth between the 3ds <-> PC..

**Reply by kanjieater** [reply]
While I got it to work once for Summon night 2, I just confirmed this didn't work for Mother 3 JPN (no intro). 

- Created save with first save point in game with oaf
- Transfered .sav to pc
- Converted with eeprom convertor
- Renamed .sav to .srm to work with retorarch's mgba core
- ran with mgba core
- game only shows New Game on home screen (but shows Load on oaf)

 I did the same thing with SN2 which was was working.

Here is the save from oaf, that still loads fine in oaf but not mgba.
[mother 3 sav.zip](https://github.com/user-attachments/files/17847691/mother.3.sav.zip)

**Reply by profi200** [reply]
When the first GBA emulators appeared the devs made a mistake and stored EEPROM saves in the wrong byte order. Until today emulators kept this format but the hardware on 3DS does use the correct byte order. While i can swap the bytes in software it requires extra RAM for another buffer plus i want to push for the correct format. What i will likely be doing is using an entirely different file extension for EEPROM saves to solve this and i hope emu devs will adapt this format/file extension eventually.

As for the Mother 3 problem that .zip file is 22 bytes. It's empty. As far as i'm aware Mother 3 uses the flash 512k save type. Not SRAM.

**Reply by kanjieater** [reply]
Thanks for the clarification. So your point about "games will try to delete your save" is that some emulators may interpret the save data in the wrong order, and this would do bad things to the save.




Sorry, not sure why that zip'd wrong. My .gba file is named like `Mother 3 (Japan).gba` , so I assumed it would get picked up by the database to get the correct format. Why would it be getting the wrong format? (as it seems like its 64KB) [Mother 3 (Japan).zip](https://github.com/user-attachments/files/17872338/Mother.3.Japan.zip)

**Reply by profi200** [reply]
The size is correct for flash 512k. If the game is patched in any way including the translations it can't be detected because of the hash mismatch. However you can put a compatible patch file along the ROM file and let oaf handle the patching at boot. It then can calculate the hash before it's patched.

Only clean dumps can be detected correctly via the database.

**Reply by kanjieater** [reply]
The md5 hash here matches the ROM, so it seems like the correct 512k is being used.  https://datomatic.no-intro.org/index.php?page=show_record&s=23&n=2347

Yet putting it through the conversion tool, it still is not being read in mGBA 🤔

**Reply by profi200** [reply]
The save you uploaded does work in mGBA for me. Get the latest version of mGBA and then reset all settings to defaults.

![working](https://github.com/user-attachments/assets/49924b9b-1242-44fa-9e58-fa962263b76e)

**Reply by kanjieater** [reply]
Of course you are correct... Embarrassing, but I thought the eeprom fixer tool had to be used to convert saves both to and from OAF. To get saves from OAF to mGBA/retroarch all you need to do is

1. Copy save to mGBA save folder
2. Rename from .sav to .srm for retroarch

And to convert a save created in Retroarch back to OAF, all you need to do is:
1. Rename it to .sav
2. run it through [this](https://exelotl.github.io/gba-eeprom-save-fix/) 
3. put it in the OAF save dir 


Thank you for your help @profi200 !!

**Reply by profi200** [reply]
It does have to be used to convert back and forth. But only for EEPROM type saves. Unless the emulator arbitrarily increases file size they are either 512 bytes or 8 KiB (if the game uses RTC then +16 bytes in some cases).

**Reply by kanjieater** [reply]
Is there any clear way to know which saves are eeprom and therefore need to be converted to & from?
